strawberry/field.py

```diff
@@ -23,129 +23,289 @@
-        self.directives = directives
+                elif getattr(argument.type, "_type_definition", False):
+                    if argument.type._type_definition.is_interface:  # type: ignore
+                        raise InvalidFieldArgument(
+                            self.base_resolver.name,
+                            argument.python_name,
+                            "Interface",
+                        )
+        # checks that only applied after origin has been determined.
+        # this would normally be checked after StrawberryField.evolve has been called.
+        if self.origin:
+            # Check that the field type is not Private
+            if is_private(self.type_annotation.resolve()):
+                raise PrivateStrawberryFieldError(
+                    self.python_name, self.origin.__name__
+                )
+            # Check that default is not set if a resolver is defined
+            if (
+                self.dataclass_options.default is not UNSET
+                and self.base_resolver is not None
+            ):
+                raise FieldWithResolverAndDefaultValueError(
+                    self.python_name, self.origin.__name__
+                )
+            # Check that default_factory is not set if a resolver is defined
+            # Note: using getattr because of this issue:
+            # https://github.com/python/mypy/issues/6910
+            if (
+                self.dataclass_options.default_factory is not UNSET  # noqa
+                and self.base_resolver is not None
+            ):
+                raise FieldWithResolverAndDefaultFactoryError(
+                    self.python_name, self.origin.__name__
+                )
 
-        self.deprecation_reason = deprecation_reason
+            assert_message = (
+                "Field must have a name by the time the schema is generated"
+            )
+            assert self.python_name is not None, assert_message
 
-    def __call__(self, resolver: _RESOLVER_TYPE) -> "StrawberryField":
-        """Add a resolver to the field"""
+    @classmethod
+    def create(
+        cls,
+        python_name: Optional[str] = None,
+        graphql_name: Optional[str] = None,
+        type_annotation: Optional[Union[StrawberryAnnotation, type]] = None,
+        origin: Optional[Union[Type, Callable, staticmethod, classmethod]] = None,
+        base_resolver: Optional[StrawberryResolver] = None,
+        is_subscription: bool = False,
+        description: Optional[str] = None,
+        deprecation_reason: Optional[str] = None,
+        permission_classes=None,  # type: ignore
+        directives: Sequence[object] = (),
+        dataclass_options: Optional[DataclassFieldSettings] = DEFAULT_FIELD_SETTINGS,
+        is_basic_field: bool = False,
+        # legacy compatibility, will be passed to dataclass_options:
+        default: object = UNSET,
+        default_factory: Union[Callable[[], Any], object] = UNSET,
```

Comment:
why is this legacy? this is (also) part of Strawberry's api 😊

Reply:
I want to remove this and let users use 
`DataclassFieldSettings` instead, this will allow to modify the fields kwargs as well as provide better typing information.

also this is not related to `StrawberryField` I want to dataclass logic from strawberry logic as much as possible.

Reply:
What do you have in mind? Should i drop the `DataclassFieldSettings`?

Reply:
yup 😊 we can declare all the params on our function, and maybe use this in future to make it cleaner; https://peps.python.org/pep-0692/